File: 2023/day01/main.c
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>

#define MAX_CHARS 100
typedef char Line[MAX_CHARS];

int part1(Line line, size_t i) {
    if (isdigit(line[i])) {
        return line[i] - '0';
    }
    return 0;
}

int part2(Line line, size_t i) {
    if (isdigit(line[i])) {
        return line[i] - '0';
    }
    Line word = {'\0'};
    size_t j = 0;
    while (line[i] != '\0') {
        word[j++] = line[i++];
        if (strcmp(word, "one") == 0) {
            return 1;
        }
        if (strcmp(word, "two") == 0) {
            return 2;
        }
        if (strcmp(word, "three") == 0) {
            return 3;
        }
        if (strcmp(word, "four") == 0) {
            return 4;
        }
        if (strcmp(word, "five") == 0) {
            return 5;
        }
        if (strcmp(word, "six") == 0) {
            return 6;
        }
        if (strcmp(word, "seven") == 0) {
            return 7;
        }
        if (strcmp(word, "eight") == 0) {
            return 8;
        }
        if (strcmp(word, "nine") == 0) {
            return 9;
        }
    }
    return 0;
}

int main(int argc, char* argv[]) {
    FILE* file = fopen("input", "r");
    if (file == NULL) {
        perror("Error opening file");
        return 1;
    }

    int (*is_digit)(Line, size_t) = part1;
    if (argc == 2 && strcmp(argv[1], "--part2") == 0) {
        is_digit = part2;
    }

    Line line = {'\0'};
    int result = 0;

    while (fgets(line, MAX_CHARS, file) != NULL) {
        printf("%s", line);
        int first;
        int second;
        size_t i = 0;
        do {
            first = is_digit(line, i++);
        } while (!first);
        for (i = 0; line[i + 1] != '\0'; ++i) {
            int value = is_digit(line, i);
            if (value) {
                second = value;
            }
        }
        printf("val: %i%i\n", first, second);
        result += first * 10 + second;
    }

    printf("%i", result);

    fclose(file);
    return 0;
}

File: 2023/day01/main.rs
fn main() {
    let file_data = include_str!("input");
    let part1: u32 = file_data
        .lines()
        .map(|line| {
            let calibrated: Vec<u32> = line.chars().filter_map(|char| char.to_digit(10)).collect();
            let first = calibrated[0];
            let last = calibrated.last().unwrap();
            first * 10 + last
        })
        .sum();
    let part2: u32 = file_data
        .lines()
        .map(|line| {
            println!("{}", line);
            let calibrated: Vec<u32> = line
                .chars()
                .enumerate()
                .filter_map(|(i, char)| {
                    char.to_digit(10).or_else(|| {
                        for end in i..=line.len() {
                            let word = &line[i..end];
                            if word == "one" {
                                return Some(1);
                            } else if word == "two" {
                                return Some(2);
                            } else if word == "three" {
                                return Some(3);
                            } else if word == "four" {
                                return Some(4);
                            } else if word == "five" {
                                return Some(5);
                            } else if word == "six" {
                                return Some(6);
                            } else if word == "seven" {
                                return Some(7);
                            } else if word == "eight" {
                                return Some(8);
                            } else if word == "nine" {
                                return Some(9);
                            }
                        }
                        None
                    })
                })
                .collect();
            let first = calibrated[0];
            let last = calibrated.last().unwrap();
            println!("{}{}", first, last);
            first * 10 + last
        })
        .sum();
    println!("part1: {}, part2: {}", part1, part2);
}
